### PR TITLE
Add Tutorial of  QAT with X86InductorQuantizer

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
## Description
This diff adds the new tutorial of `PyTorch 2 Export Quantization-Aware Training (QAT) with X86 Backend through Inductor`.

## Checklist

- [x] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [x] Only one issue is addressed in this pull request
- [x] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnecessary issues are included into this pull request.

### CODE REVIEW

**Comment by leslie-fang-intel** [comment]
cc @jgong5 @jerryzh168 @svekars please help to take a look of this PR.

**Review by jgong5 — APPROVED** [review]

File: prototype_source/pt2e_quant_qat_x86_inductor.rst [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+-  `PyTorch 2 Export Post Training Quantization with X86 Backend through Inductor tutorial <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_x86_inductor.html>`_
+-  `TorchInductor and torch.compile concepts in PyTorch <https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html>`_
+
+
+This tutorial demonstrates the process of performing PT2 export quantization-aware training (QAT) on X86 CPU
+with X86InductorQuantizer, and subsequently lowering the quantized model into Inductor.
+For more comprehensive details about PyTorch 2 Export Quantization-Aware Training in general, please refer to the
+dedicated tutorial on `PyTorch 2 Export Quantization-Aware Training <https://pytorch.org/tutorials/prototype/pt2e_quant_qat.html>`_.
+For a deeper understanding of X86InductorQuantizer, please consult the tutorial of
+`PyTorch 2 Export Post Training Quantization with X86 Backend through Inductor <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_x86_inductor.html>`_.
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_x86_inductor.rst:19]:
```suggestion
This tutorial demonstrates the process of performing PT2 export Quantization-Aware Training (QAT) on X86 CPU
using X86InductorQuantizer and subsequently lowering the quantized model into Inductor.
For a more in-depth understanding of PT2 Export Quantization-Aware Training, we recommend referring to the dedicated `PyTorch 2 Export Quantization-Aware Training <https://pytorch.org/tutorials/prototype/pt2e_quant_qat.html>`_.
To gain a deeper insight into X86InductorQuantizer, please see the tutorial of
`PyTorch 2 Export Post Training Quantization with X86 Backend through Inductor <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_x86_inductor.html>`_.
```

File: prototype_source/pt2e_quant_qat_x86_inductor.rst [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+The PyTorch 2 Export QAT flow looks like the following—it is similar
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_x86_inductor.rst:21]:
```suggestion
The PyTorch 2 Export QAT flow is largely similar to the
```

File: prototype_source/pt2e_quant_qat_x86_inductor.rst [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+to the post training quantization (PTQ) flow for the most part:
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_x86_inductor.rst:22]:
```suggestion
to the post-training quantization (PTQ) flow for the most part:
```
If post-training fails the spellcheck, add "post-training" to en-wordlist.txt

File: prototype_source/pt2e_quant_qat_x86_inductor.rst [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  # move the quantized model to eval mode, equivalent to `m.eval()`
+  torch.ao.quantization.move_exported_model_to_eval(converted_model)
+
+  # Lower the model into Inductor
+  with torch.no_grad():
+    optimized_model = torch.compile(converted_model)
+    _ = optimized_model(*example_inputs)
+
+Please note that since the Inductor ``freeze`` feature does not turn on by default yet, need to run example code with ``TORCHINDUCTOR_FREEZING=1``.
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_x86_inductor.rst:72]:
```suggestion
Please note that the Inductor ``freeze`` feature is not enabled by default. To use this feature, you need to run example code with ``TORCHINDUCTOR_FREEZING=1``.
```

**Review by svekars — COMMENTED** [review]
Thank you, @leslie-fang-intel! Just a couple editorial suggestions.

**Comment by jerryzh168** [comment]
should we incorporate this into the PTQ tutorial to reduce the number of tutorials?

Maybe we should do the same for XNNPACKQuantizer as well, cc @andrewor14

**Review by svekars — COMMENTED** [review]
Also, please add to https://github.com/pytorch/tutorials/blob/main/prototype_source/prototype_index.rst

**Comment by andrewor14** [comment]
Hi @leslie-fang-intel, thanks for working on this. I agree with @jerryzh168 that we should merge this with the X86Inductor PTQ tutorial. What you're adding here seems very similar to the existing PT2 export QAT tutorial except for maybe these 3 lines:
```
  quantizer = X86InductorQuantizer()
  quantizer.set_global(xiq.get_default_x86_inductor_quantization_config(is_qat=True))
  prepared_model = prepare_qat_pt2e(exported_model, quantizer)
```
so I feel it's not worth making a brand new tutorial just for this use case. What do you think?

**Comment by leslie-fang-intel** [comment]
@jerryzh168 @andrewor14 Thanks for the comment. Merge the `PTQ` and `QAT` for `X86InductorQuantizer` into one tutorial. Please kindly help to take a look again.

**Comment by leslie-fang-intel** [comment]
@svekars I am afraid the changes in this PR may have conflict with https://github.com/pytorch/tutorials/pull/2716. Should we merge these 2 PRs together to facilitate the review or merge process? Or do you prefer to merge https://github.com/pytorch/tutorials/pull/2716 firstly then we rebase current PR after it.

**Comment by svekars** [comment]
> @svekars I am afraid the changes in this PR may have conflict with #2716. Should we merge these 2 PRs together to facilitate the review or merge process? Or do you prefer to merge #2716 firstly then we rebase current PR after it.

I'm good either way, if you have bandwidth you can merge these two in one PR and close the one we don't need to merge. But also happy with the rebase plan.

File: prototype_source/pt2e_quant_x86_inductor.rst [hunk 5]
```diff
@@ -15,14 +16,15 @@ Introduction
 This tutorial introduces the steps for utilizing the PyTorch 2 Export Quantization flow to generate a quantized model customized
 for the x86 inductor backend and explains how to lower the quantized model into the inductor.
 
-The new quantization 2 flow uses the PT2 Export to capture the model into a graph and perform quantization transformations on top of the ATen graph. This approach is expected to have significantly higher model coverage, better programmability, and a simplified UX.
+The new quantization 2 flow uses the PT2 Export to capture the model into a graph and perform quantization transformations on top of the ATen graph.
```
Comment (jerryzh168) [inline prototype_source/pt2e_quant_x86_inductor.rst:19]: nit: we don't use `quantization 2 flow`, please use `pytorch 2 export quantization flow` instead

File: prototype_source/pt2e_quant_x86_inductor.rst [hunk 6]
```diff
@@ -154,6 +176,7 @@ Finally, we will convert the calibrated Model to a quantized Model. ``convert_pt
 ::
 
     converted_model = convert_pt2e(prepared_model)
+    torch.ao.quantization.move_exported_model_to_eval(converted_model)
```
Comment (jerryzh168) [inline prototype_source/pt2e_quant_x86_inductor.rst:179]: is this needed for both PTQ and QAT?

Comment (jerryzh168) [reply prototype_source/pt2e_quant_x86_inductor.rst:179]: I feel we could also have a separate section for QAT, like what we did in https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html#quantization-aware-training

File: prototype_source/pt2e_quant_x86_inductor.rst (hunk 5 above)
Comment (leslie-fang-intel) [reply prototype_source/pt2e_quant_x86_inductor.rst:19]: Thanks and changed.

File: prototype_source/pt2e_quant_x86_inductor.rst (hunk 6 above)
Comment (leslie-fang-intel) [reply prototype_source/pt2e_quant_x86_inductor.rst]: Thanks for the comment. Yean, I think it will be much clearer for reader. Move the QAT into a separate section. Please kindly help to take a look again @jerryzh168.

**Comment by leslie-fang-intel** [comment]
Hi @svekars @jerryzh168, I have rebased to fix the conflict. Please kindly help to take a look again.

**Review by svekars — APPROVED** [review]
